test/components/views/settings/discovery/EmailAddresses-test.tsx

```diff
@@ -15,26 +15,137 @@ limitations under the License.
 */
 
 import React from "react";
-import { render, screen } from "@testing-library/react";
+import { fireEvent, render, screen } from "@testing-library/react";
 import { IThreepid, ThreepidMedium } from "matrix-js-sdk/src/@types/threepids";
+import { IRequestTokenResponse } from "matrix-js-sdk/src/client";
+import { MatrixError } from "matrix-js-sdk/src/http-api";
 
+import { UserFriendlyError } from "../../../../../src/languageHandler";
 import { EmailAddress } from "../../../../../src/components/views/settings/discovery/EmailAddresses";
+import { clearAllModals, getMockClientWithEventEmitter } from "../../../../test-utils";
+
+const mockGetAccessToken = jest.fn().mockResolvedValue("getAccessToken");
+jest.mock("../../../../../src/IdentityAuthClient", () =>
+    jest.fn().mockImplementation(() => ({
+        getAccessToken: mockGetAccessToken,
+    })),
+);
+
+const emailThreepidFixture: IThreepid = {
+    medium: ThreepidMedium.Email,
+    address: "foo@bar.com",
+    validated_at: 12345,
+    added_at: 12342,
+    bound: false,
+};
 
 describe("<EmailAddress/>", () => {
+    const mockClient = getMockClientWithEventEmitter({
+        getIdentityServerUrl: jest.fn().mockReturnValue("https://fake-identity-server"),
+        generateClientSecret: jest.fn(),
+        doesServerSupportSeparateAddAndBind: jest.fn(),
+        requestEmailToken: jest.fn(),
+        bindThreePid: jest.fn(),
+    });
+
+    beforeEach(() => {
+        jest.useFakeTimers();
+    });
+
+    afterEach(async () => {
+        jest.useRealTimers();
+        await clearAllModals();
+    });
+
     it("should track props.email.bound changes", async () => {
-        const email: IThreepid = {
-            medium: ThreepidMedium.Email,
-            address: "foo@bar.com",
-            validated_at: 12345,
-            added_at: 12342,
-            bound: false,
-        };
-
-        const { rerender } = render(<EmailAddress email={email} />);
+        const { rerender } = render(<EmailAddress email={emailThreepidFixture} />);
         await screen.findByText("Share");
 
-        email.bound = true;
-        rerender(<EmailAddress email={{ ...email }} />);
+        rerender(
+            <EmailAddress
+                email={{
+                    ...emailThreepidFixture,
+                    bound: true,
+                }}
+            />,
+        );
         await screen.findByText("Revoke");
     });
+
+    describe("Email verification share phase", () => {
```

Comment:
I've added a few tests just to sanity check that the new `UserFriendlyError` are doing things as expected.

I didn't add test coverage for all of these methods where we changed the error handling as that requires writing tests for these entire components as they were barely tested before. Which leads us to the SonarCloud coverage failures.